http: edit page (#13507)
tldr-pages/tldr: pages/common/http.md

# http

> > HTTPie: designed for testing, debugging, and generally interacting with APIs & HTTP servers.

Reviewer: ```suggestion
> HTTPie: an HTTP client designed for testing, debugging, and generally interacting with APIs & HTTP servers.
```

> More information: <https://httpie.io/docs/cli/usage>.

- Make a simple GET request (shows response header and content):

`http {{https://example.org}}`

> - Print specific output content (`H`: request headers, `B`: request body, `h`: response headers, `b`: response body, `m`: response metadata):

Reviewer 2: ```suggestion
- Print specific output content (request [H]eaders, request [B]ody, response [h]eaders, response [b]ody, response [m]etadata):
```

Reviewer: This syntax is used only for highlighting short options.
https://github.com/tldr-pages/tldr/blob/main/contributing-guides/style-guide.md#short-option-mnemonics

Reviewer 2: But why dont use this syntax in this case?

Author: Because it is not an option itself, but a value to the option. The syntax is meant for short options.

Reviewer: If we start using this for other things, it'll quickly get confusing. How will you know whether it denotes a short option or something else?

`http --print {{H|B|h|b|m|Hh|Hhb|...}} {{https://example.com}}`

> - Specify HTTP method when sending a request:

Reviewer: ```suggestion
- Specify the HTTP method when sending a request:
```

`http {{GET|POST|HEAD|PUT|PATCH|DELETE|...}} {{https://example.com}}`

- Specify additional headers in a request:

`http {{https://example.com}} {{'User-Agent: Mozilla/5.0' 'Accept-Encoding: gzip'}}`

- Authenticate to a server using different authentication methods:

`http --auth {{username:password|token}} --auth-type {{basic|digest|bearer}} {{GET|POST|...}} {{https://example.com/auth}}`

- Construct a request but do not send it (similar to a dry-run):

`http --offline {{GET|DELETE|...}} {{https://example.com}}`

> - Use named sessions for persistent requests between requests:

Reviewer: From the man page:
> --session SESSION_NAME_OR_PATH
>
> Create, or reuse and update a session. Within a session, custom headers, auth credential, as well as any cookies sent by the server persist between requests.
>
> Session files are stored in:
> `[HTTPIE_CONFIG_DIR]/<HOST>/<SESSION_NAME>.json`.
```suggestion
- Use named sessions for persistent custom headers, auth credentials and cookies:
```

> `http --session={{path/to/logged-in-session.json}} {{--auth username:password https://example.com/auth API-KEY:xxx}}`

Reviewer: ```suggestion
`http --session {{session_name|path/to/session.json}} --auth {{username}}:{{password}} https://example.com/auth {{API-KEY:xxx}}`
```

Author: I would like to stick to the original example. Using session does not necessarily require use the of `--auth`. An unauthenticated session can also have relevant session information. Hence I enclosed the whole `{{--auth username:password https://example.com/auth API-KEY:xxx}}` in double braces.

Reviewer: Ok, what about this? `--session {{name}}` stores the session file in the default location. 
```suggestion
`http --session {{session_name|path/to/session.json}} {{--auth username:password https://example.com/auth API-KEY:xxx}}`
```

> - Upload a file to a form (example below assumes form field is `<input type="file" name="cv" />`):

Reviewer: ```suggestion
- Upload a file to a form (the example below assumes form field is `<input type="file" name="cv" />`):
```
